File: packages/platform/src/ipc/address.ts
/**
 * 发布地址库 IPC 封装 — 全局随机地址池 + 个人地址库 CRUD。
 *
 * 后端：壳层 `commands/address.rs`（InMemoryAddressStore + app::publish::AddressService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { call } from "./invoke";

/** 地址类型（与 Rust `AddressType` 对齐）。 */
export type AddressType = "global" | "personal";

/** 发布地址（与 Rust `PublishAddress` 对齐核心字段）。 */
export interface PublishAddress {
  id: number;
  owner_id: number;
  address_type: AddressType;
  address: string;
  name: string;
  search_keyword: string;
  expected_text: string | null;
  weight: number;
  sort_order: number;
  is_enabled: boolean;
  use_count: number;
  remark: string | null;
  created_at: string | null;
  updated_at: string | null;
}

const OWNER_ID = 1; // 桌面单用户；多用户时由登录态注入

/** 分页查询地址。 */
export function addressList(query: {
  page: number;
  page_size: number;
  keyword?: string;
  address_type?: AddressType;
}): Promise<[PublishAddress[], number]> {
  return call<[PublishAddress[], number]>("address_list", {
    request: {
      owner_id: OWNER_ID,
      page: query.page,
      page_size: query.page_size,
      keyword: query.keyword ?? "",
      address_type: query.address_type ?? "",
    },
  });
}

/** 新建地址。 */
export function addressCreate(
  address: Omit<PublishAddress, "id" | "owner_id">,
): Promise<PublishAddress> {
  return call<PublishAddress>("address_create", {
    ownerId: OWNER_ID,
    address: { ...address, id: 0, owner_id: OWNER_ID },
  });
}

/** 更新地址。 */
export function addressUpdate(address: PublishAddress): Promise<void> {
  return call<void>("address_update", { ownerId: OWNER_ID, address });
}

/** 删除地址。 */
export function addressDelete(addressId: number): Promise<void> {
  return call<void>("address_delete", {
    request: { owner_id: OWNER_ID, address_id: addressId },
  });
}

/** 批量删除地址（返回实际删除数量）。 */
export function addressBatchDelete(addressIds: number[]): Promise<number> {
  return call<number>("address_batch_delete", {
    request: { owner_id: OWNER_ID, address_ids: addressIds },
  });
}

File: apps/desktop/src-tauri/src/platforms/xianyu/ipc/address.rs
//! 发布地址 Tauri commands — 地址 CRUD + 批量删除。

use crate::platforms::xianyu::persist::InMemoryAddressStore;
use app::publish::{AddressQuery, AddressService, PublishAddress};
use common;
use serde::Deserialize;
use std::sync::Arc;
use tauri::State;

/// 地址服务句柄（setup 时注册到 Tauri 状态）。
pub struct AddressHandle {
    pub store: Arc<InMemoryAddressStore>,
}

#[derive(Debug, Deserialize)]
pub struct AddressListRequest {
    pub owner_id: i64,
    pub page: u32,
    pub page_size: u32,
    #[serde(default)]
    pub keyword: String,
    /// global / personal / 空。
    #[serde(default)]
    pub address_type: String,
}

#[derive(Debug, Deserialize)]
pub struct AddressDeleteRequest {
    pub owner_id: i64,
    pub address_id: i64,
}

#[derive(Debug, Deserialize)]
pub struct AddressBatchDeleteRequest {
    pub owner_id: i64,
    pub address_ids: Vec<i64>,
}

#[tauri::command]
pub fn address_list(
    state: State<'_, AddressHandle>,
    request: AddressListRequest,
) -> common::OpenDeskResult<(Vec<PublishAddress>, u32)> {
    let service = AddressService::new(state.store.as_ref());
    let query = AddressQuery {
        page: request.page,
        page_size: request.page_size,
        keyword: request.keyword,
        address_type: request.address_type,
    };
    service.list(request.owner_id, &query)
}

#[tauri::command]
pub fn address_create(
    state: State<'_, AddressHandle>,
    owner_id: i64,
    address: PublishAddress,
) -> common::OpenDeskResult<PublishAddress> {
    let service = AddressService::new(state.store.as_ref());
    service
        .create(owner_id, address)
        .map_err(common::OpenDeskError::wrap)
}

#[tauri::command]
pub fn address_update(
    state: State<'_, AddressHandle>,
    owner_id: i64,
    address: PublishAddress,
) -> common::OpenDeskResult<()> {
    let service = AddressService::new(state.store.as_ref());
    service
        .update(owner_id, &address)
        .map_err(common::OpenDeskError::wrap)
}

#[tauri::command]
pub fn address_delete(
    state: State<'_, AddressHandle>,
    request: AddressDeleteRequest,
) -> common::OpenDeskResult<()> {
    let service = AddressService::new(state.store.as_ref());
    service.delete(request.owner_id, request.address_id)
}

#[tauri::command]
pub fn address_batch_delete(
    state: State<'_, AddressHandle>,
    request: AddressBatchDeleteRequest,
) -> common::OpenDeskResult<usize> {
    let service = AddressService::new(state.store.as_ref());
    service.batch_delete(request.owner_id, &request.address_ids)
}
